storage/src/Database.cpp

```diff
@@ -129,13 +130,101 @@ void Database::open_and_prepare(const std::string& db_path) {
     if (!get_stmt)
         throw std::runtime_error("could not prepare get statement");
 
+    get_row_count_stmt = prepare_statement("SELECT count(*) FROM `Data`;");
+    if (!get_row_count_stmt)
+        throw std::runtime_error("could not prepare row count statement");
+
+    get_by_index_stmt = prepare_statement("SELECT * FROM `Data` LIMIT ?, 1;");
+    if (!get_by_index_stmt)
+        throw std::runtime_error("could not prepare get by index statement");
+
     delete_expired_stmt =
         prepare_statement("DELETE FROM `Data` WHERE `TimeExpires` <= ?");
     if (!delete_expired_stmt)
         throw std::runtime_error(
             "could not prepare 'delete expired' statement");
 }
 
+bool Database::get_message_count(uint64_t& count) {
+
+    int rc;
+    bool success = false;
+    while (true) {
+        rc = sqlite3_step(get_row_count_stmt);
+        if (rc == SQLITE_BUSY) {
+            continue;
+        } else if (rc == SQLITE_DONE) {
+            break;
+        } else if (rc == SQLITE_ROW) {
+            count = sqlite3_column_int64(get_row_count_stmt, 0);
+            success = true;
+        } else {
+            BOOST_LOG_TRIVIAL(error) << "Could not execute `count` db statement";
+            break;
+        }
+    }
+
+    rc = sqlite3_reset(get_by_index_stmt);
+    if (rc != SQLITE_OK) {
+        BOOST_LOG_TRIVIAL(error) << "sqlite reset error: " << rc;
+        success = false;
+    }
+
+    return success;
+}
+
+/// Extract item from the result of a successfull select statement execution
+static Item extract_item(sqlite3_stmt* stmt) {
+
+    Item item;
+
+    // "If the SQL statement does not currently point to a valid row, or if the
+    // column index is out of range, the result is undefined"
+    item.hash = std::string((const char*)sqlite3_column_text(stmt, 0));
+    item.pub_key = std::string((const char*)sqlite3_column_text(stmt, 1));
+    item.ttl = sqlite3_column_int64(stmt, 2);
+    item.timestamp = sqlite3_column_int64(stmt, 3);
+    item.expiration_timestamp = sqlite3_column_int64(stmt, 4);
+    item.nonce = std::string((const char*)sqlite3_column_text(stmt, 5));
+    item.data = std::string((char*)sqlite3_column_blob(stmt, 6),
+                            sqlite3_column_bytes(stmt, 6));
+    return item;
+}
+
+bool Database::retrieve_by_index(uint64_t index, Item& item) {
```

Comment:
What will be the purpose of this function?
(The index can't be relied upon to identify messages across snodes)

Reply:
If I know that there is 100 messages, I can generate a random number in [0; 100) and request it using this method.

Reply:
My question was more related to the purpose of the function in regard to the  storage tests :)

Reply:
I will pick a random message and send its owner's pubkey to the node being tested. That node will need to find it in the database and return the message/hash.

Reply:
Ah ok all good then !